### DISCUSSION TITLE
Print config without token

### DISCUSSION THREAD

**Discussion by ph-fritsche** [renovatebot/renovate#8882] [Request Help]
**What would you like Renovate to be able to do?**

Print resolved config on any machine without GitHub personal access token.

Currently `$ renovate --print-config` fails with `FATAL: You must configure a GitHub personal access token`

**Did you already have any implementation ideas?**

Skip the check for personal access token when `--print-config` is `true`.

**Workaround**

When environment variable `LOG_LEVEL` is set to `debug`, the output contains the resolved config as debug information.

**Comment by rarkins** [comment]
`--print-config` is for printing the *repository* config, not the bot config. You never get to the repository config without a token. To view bot config, you must do so with log level.

**Reply by ph-fritsche** [reply]
Thanks for the quick reply.

So if there is no support for resolving configs for config authors aside from debugging, I've got a follow-up question:
Is the only way to verify a (shared) config to run it on a burner-repo and inspect the results?
(And then later copy it to the real config)

**Reply by rarkins** [reply]
> So if there is no support for resolving configs for config authors

I don't understand what you mean by that.

BTW if this is a public repo, have you tried simply giving a dummy token?

Also. have you seen the `renovate-config-validator` tool which installs alongside `renovate`?

**Reply by ph-fritsche** [reply]
Yes, already running the validator in CI. Thanks for the hint anyway :)
I stumbled over it when inspecting how you guys implement shared configs yourself.

The problem occurring to me after reading the docs and inspecting some setups:
Once someone has a shared config in place, how can anyone extend that safely?
Especially if there are any `automerge` rules in effect?

Before wiring everything together I'd like to put some verification in place, so that changes could be implemented with some confidence later - without rereading every part of the docs and configs.

The hint on the dummy token is great. So `--dry-run` on a public repo with a dummy token should work?
So could I set up public example repo, set up a new config somewhere (local/ci) and let `renovate --dry-run` run on that example repo with that new config and a dummy token?

**Reply by rarkins** [reply]
Yes, I think if you run dry run with a dummy token it should get all the way through, including printing the config. Naturally the preset config you're extending would need to be publicly accessible.

**Reply by ph-fritsche** [reply]
Sadly this does not work with a dummy token.
The CLI tries to access `GET https://api.github.com/user` which fails with `401 Unauthorized` `Bad Credentials`.

So if I set up a CI job running `renovate` on push requests with a token that lacks the scopes needed for doing actual changes, could renovate pick up the config file from the PR HEAD so that I could run renovate with that new config there?

**Reply by rarkins** [reply]
If you have a read only token from a dummy *user* account then maybe. But can you also try again with a dummy token but specifying a dummy `username` too? That might shortcut the user API call

**Reply by ph-fritsche** [reply]
`--username <string>                                Username for authentication. Currently Bitbucket only.`
Tried it anyway. No effect. :(

**Reply by rarkins** [reply]
I use it in config.js when running using app tokens (which also can't access /user)

**Reply by ph-fritsche** [reply]
Tried per config file. Same result. `initPlatform()` also requires `gitAuthor` to be set. Then it does skip accessing `/user`.
But then comes the access on `POST https://api.github.com/graphql`. That fails with `401 Unauthorized` of course.
And omitting the token leads back to the starting point.

On second thought I guess a dry run without a user token would fail due to access rate limits anyway.

**Reply by ph-fritsche** [reply]
So here's an update:
I've set up a workflow running `renovate --dry-run` on both the base and head of a PR and committing the logs so that one can view the diff.
Have a look at one of the logs [here](https://github.com/ph-fritsche/renovate-config/commit/f66b8fcd25b128cdd07cc591bc7c8524145c2103).

That is a step forward in providing information about how a change to the config would alter the expected changes when running `renovate`.

The provided information about the combined config looks good to me. This should give any maintainer more confidence than resolving it manually.

The dry run log about the PRs to be generated and the updated dependencies is a lot easier to grasp with the diff highlighted.
But there is still room for improvement.

I think the dry run log should contain `INFO` about the changes that will be proposed in the PRs.
The `DRY-RUN: Would commit files to branch renovate/...` should be extended by information like `newVersion` and `newValue` (as provided for PHP packages under `DEBUG: packageFiles with updates`) and the commit message.
Maybe the diff of the proposed commits could be logged on `INFO` level. This way the output would be similar for the different managers and one could change the `log_level` to reduce the noise.

What do you think?

**Reply by rarkins** [reply]
I'm fine with enhancements to the dry-run verboseness
